server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java

```diff
@@ -90,75 +118,116 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
+        if (existingConnectionRecheck != null) {
+            ListenableFuture<Transport.Connection> future = pendingConnections.remove(node);
             assert future == currentListener : "Listener in pending map is different than the expected listener";
             connectingRefCounter.decRef();
-            future.onResponse(null);
+            future.onResponse(existingConnectionRecheck);
             return;
         }
 
         final RunOnce releaseOnce = new RunOnce(connectingRefCounter::decRef);
-        internalOpenConnection(node, resolvedProfile, ActionListener.wrap(conn -> {
-            connectionValidator.validate(conn, resolvedProfile, ActionListener.wrap(
+        internalOpenConnection(node, resolvedProfile, ActionListener.wrap(
+            conn -> connectionValidator.validate(conn, resolvedProfile, ActionListener.runAfter(ActionListener.wrap(
                 ignored -> {
                     assert Transports.assertNotTransportThread("connection validator success");
                     try {
                         if (connectedNodes.putIfAbsent(node, conn) != null) {
-                            assert false : "redundant conection to " + node;
-                            logger.debug("existing connection to node [{}], closing new redundant connection", node);
+                            assert false : "redundant connection to " + node;
+                            logger.warn("existing connection to node [{}], closing new redundant connection", node);
                             IOUtils.closeWhileHandlingException(conn);
                         } else {
                             logger.debug("connected to node [{}]", node);
                             try {
                                 connectionListener.onNodeConnected(node, conn);
                             } finally {
-                                final Transport.Connection finalConnection = conn;
                                 conn.addCloseListener(ActionListener.wrap(() -> {
-                                    logger.trace("unregistering {} after connection close and marking as disconnected", node);
-                                    connectedNodes.remove(node, finalConnection);
+                                    connectedNodes.remove(node, conn);
                                     connectionListener.onNodeDisconnected(node, conn);
+                                    conn.onRemoved();
+                                }));
+
+                                conn.addCloseListener(ActionListener.wrap(() -> {
+                                    if (connectingRefCounter.hasReferences() == false) {
+                                        logger.trace("connection manager shut down, closing transport connection to [{}]", node);
+                                    } else if (conn.hasReferences()) {
+                                        logger.info("transport connection to [{}] closed by remote", node.descriptionWithoutAttributes());
```

Comment:
This also emits logs for connections to nodes in remote clusters. I think that's useful to see, we investigate cases that turn out to be flaky cross-cluster connections at a nonzero rate.